### PR TITLE
fix groupcalls

### PR DESCRIPTION
**Pull request by sudden6**
change the condition for when the audiocall button is enabled

affects #2989

### CODE REVIEW

**Comment by zetok** [comment]
Well, that does fix #2989 for me.

Now, the problem that I have is that client's interface hardlocks when I'm trying to end the call. 

```
[16:22:47.235] src/audio/audio.cpp:292 : Debug: Opened audio output "Built-in Audio Analog Stereo"
[New LWP 5818]
[16:22:47.236] src/audio/audio.cpp:561 : Debug: Audio source 2 created. Sources active: 1
[16:22:47.236] src/audio/audio.cpp:181 : Debug: Subscribed to audio input device [ 2 subscriptions ]
[16:22:47.236] src/audio/audio.cpp:197 : Debug: Unsubscribed from audio input device [ 1 subscriptions left ]
[16:22:47.238] src/video/groupnetcamview.cpp:42 : Debug: Created expanding?  false
[16:22:47.238] src/video/groupnetcamview.cpp:42 : Debug: Created expanding?  true
[16:22:47.241] src/video/groupnetcamview.cpp:42 : Debug: Created expanding?  true
[16:22:53.196] src/core/coreav.cpp:388 : Debug: "Leaving group call 0"
[16:25:40.882] src/core/core.cpp:522 : Debug: "Group namelist change 0:1 1"
[16:26:17.182] src/core/corefile.cpp:285 : Debug: "Received avatar request 2:65536, reject, since we have it in cache."
```

**Comment by anoadragon453** [comment]
This does not always fix the problem. I'm getting audio button enabled on groupchats intermittently now. When it is enabled I have the same issue as @zetok above where the client crashes upon ending the call with no useful debug information.

Additionally I noticed when a user leaves a chat, their avatar does not disappear in the avatar/visual-overview area, though that can be reserved to another issue. 

Edit: Huh, when I exited the Windows client, the audio button became enabled. Windows is not on this PR commit, but a couple back. The Linux are of course.

I also got a `src/core/core.cpp:955 : Warning: getGroupPeerName: Unknown error` message.

---

After a bit more testing if I only add clients that are on this commit then the call enabled button seems to work properly.

**Comment by anoadragon453** [comment]
While this enables the call button, calls are still not functioning correctly, the other user does not even get an indication that someone has initiated a group call.

Step in the right direction though.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 49c8034fdf0afac4ab4522e08f20d11ab87c9382
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 487e6f597349c46dab25c4479052ca3b527f3103
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 987cb16cc8bb4a204e11d752a6794142de3217ad
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at
https://github.com/tux3/qTox/blob/master/CONTRIBUTING.md#commit-message-format

If this message was triggered by missing `(<scope>)` simply ignore it.

---
This message was auto-generated by https://gitcop.com

File: src/audio/audio.cpp [hunk 1]
```diff
@@ -368,6 +368,35 @@ void Audio::playGroupAudioQueued(void*,int group, int peer, const int16_t* data,
...
+{
+    assert(QThread::currentThread() == audioThread);
+    QMutexLocker locker(&audioLock);
+
+    if (!CoreAV::groupCalls.contains(group))
+        return;
+
+    ToxGroupCall& call = CoreAV::groupCalls[group];
+
+    if (call.inactive || call.muteVol)
```
Comment (antis81) [inline src/audio/audio.cpp]: I'll copy the `call.inactive` into #3183, thus it does the same thing - just the right way. :)

File: src/audio/audio.cpp [hunk 2]
```diff
@@ -368,6 +368,35 @@ void Audio::playGroupAudioQueued(void*,int group, int peer, const int16_t* data,
...
+        return;
+
+    qreal volume = 0.;
+    int bufsize = samples * 2 * channels;
+    for (int i = 0; i < bufsize; ++i)
+        volume += abs(data[i]);
+
+    emit groupAudioPlayed(group, peer, volume / bufsize);
```
Comment (antis81) [inline src/audio/audio.cpp]: Hm… not sure if we actually need this event anymore - to me it looks wrong. I'll do some more research and see what it was actually used for.

Comment (sudden6) [inline src/audio/audio.cpp]: it was probably intended to indicate who is currently speaking.

**Comment by sudden6** [comment]
@antis81 With these commits it actually works! Great work :+1:

File: src/core/coreav.cpp [hunk 3]
```diff
@@ -364,6 +364,49 @@ void CoreAV::volMuteToggle(uint32_t callId)
...
+        qWarning() << "Audio for invalid group id" << group
+                   << "will not be played back.";
+        return;
+    }
+
+    ToxGroupCall& call = cav->groupCalls[group];
+
+    emit c->groupPeerAudioPlaying(group, peer);
+
+    if (call.muteVol)
```
Comment (antis81) [inline src/core/coreav.cpp]: You need to probably add the `call.inactive` here, but not sure.

**Comment by antis81** [comment]
Looks like either a commit was not signed or there is a typo in the travis script. It tells:

``` bash
Merge unsigned commits
refractor: abstract HTML out of `new issue` GitHub link
Update for the German translation file
Prevent regex capture from extended past line boundaries.
Add direct link with template for submitting new issues

Above ↑ commits don't conform to commit message format:
https://github.com/tux3/qTox/blob/master/CONTRIBUTING.md#commit-message-format

Pls fix.
```

It says `refractor` -> should be `refactor`. Anyway, this looks a bit strange…

**Comment by sudden6** [comment]
this is probably because this branch was started before new commit format and not yet rebased.

**Comment by sudden6** [comment]
@antis81 am I seeing this right that toxcore keeps sending the audio data of a groupchat to all peers, and there's no way of unsubscribing from the audio stream? Because this would be a huge bandwith waste.

**Comment by TheNain38** [comment]
@sudden6 Normally, there are a "text groupchat" and an "audio groupchat" which are joined when joining a groupchat with both in it, instead you could just leave the audio group, and rejoin it when you actually enter the call? (If there are no way to unsubscribe from the audio stream)

**Comment by sudden6** [comment]
I saw functions to make an av groupchat from a text groupchat, but not vice versa. How do other clients handle this?

**Comment by zetok** [comment]
@sudden6 

> I saw functions to make an av groupchat from a text groupchat, but not vice versa.

Err? AFAIK there are no functions like that – i.e. you specify groupchat type at the time of creation, and go from there.

> How do other clients handle this?

Well, time ago, Toxic was the only client that allowed users to choose which groupchat type should be started, i.e. audio or text type. What other clients did, is defaulting to audio groupchats, and "forgetting" about text-only groupchats.

As for bandwidth - IIRC one simply doesn't send audio data if there is no need.

# 

@antis81 
Well, yeah, due to how travis works, it kinda will include commits that are later-on (after PR creation) included on master if in meanwhile PR gets an update.

E.g. with

```
 A---B---C  master
  \
   D-------E  PR
```

once `E` commit gets pushed to PR, travis gets all commits `A, B, C, D, E`, even if PR was based on `A`.
And yeah, there were a few commits in meanwhile before verification script was introduced: #3137.

**Comment by antis81** [comment]
@zetok Ah ok, thanks for clarification. Seems like this PR is rebased now correctly on master and everyone can live happily ever after. :)

@sudden6 

> am I seeing this right that toxcore keeps sending the audio data of a groupchat to all peers, and there's no way of unsubscribing from the audio stream? Because this would be a huge bandwith waste.

At this very point, this is the wrong question. You ask for the goals we want to reach with qTox, which  goes far beyond the purpose of this little PR.

**Comment by sudden6** [comment]
@zetok 

> > I saw functions to make an av groupchat from a text groupchat, but not vice versa.
> 
> Err? AFAIK there are no functions like that – i.e. you specify groupchat type at the time of creation, and go from there.

Sorry, I confused some toxcore internal function with the public API.

@antis81 

> At this very point, this is the wrong question. You ask for the goals we want to reach with qTox, which goes far beyond the purpose of this little PR.

Probably, yes. We can't change this anyway, because toxcore doesn't provide a way to controll this.

**Comment by zetok** [comment]
Rebased commits are on master, thus closing.

**Pull request closed without merging**